File: android/app/src/main/kotlin/jp/bizen/flutter/practice/channel/platform/PlatformText.kt
package jp.bizen.flutter.practice.channel.platform

import android.app.Activity
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import jp.bizen.flutter.practice.channel.ui.textinput.TextInputActivity

class PlatformText private constructor(private val channel: MethodChannel) {
    companion object {
        lateinit var instance: PlatformText
        private const val TEXT_CHANNEL = "channel/text"
        private const val METHOD_LAUNCH_TEXT_SCREEN = "LaunchTextScreen"
        private const val METHOD_SEND_TEXT = "SendText"

        fun setup(rootScreen: Activity, messenger: BinaryMessenger) {
            val textChannel = MethodChannel(messenger, TEXT_CHANNEL)
            instance = PlatformText(textChannel)
            textChannel.setMethodCallHandler { call, result ->
                instance.handler(rootScreen, call, result)
            }
        }
    }

    private fun handler(rootScreen: Activity, call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            METHOD_LAUNCH_TEXT_SCREEN -> {
                TextInputActivity.createIntent(rootScreen)
                result.success(null)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    fun sendText(text: String) {
        channel.invokeMethod(METHOD_SEND_TEXT, text)
    }
}

File: ios/Runner/Platform/PlatformText.swift
import UIKit
import Flutter

class PlatformText {
    static var instance: PlatformText!
    
    private static let TEXT_CHANNEL = "channel/text"
    private static let METHOD_LAUNCH_TEXT_SCREEN = "LaunchTextScreen"
    private static let METHOD_SEND_TEXT = "SendText"
    private let channel: FlutterMethodChannel
    
    private init(_ channel: FlutterMethodChannel) {
        self.channel = channel
    }
    
    static func setup(rootScreen: FlutterViewController, _ messenger: FlutterBinaryMessenger){
        let channel = FlutterMethodChannel(name: TEXT_CHANNEL, binaryMessenger: messenger)
        instance = PlatformText(channel)
        channel.setMethodCallHandler { methodCall, result in
            instance.handler(rootScreen: rootScreen, call: methodCall, result: result)
        }
    }
    
    private func handler(rootScreen: FlutterViewController, call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case PlatformText.METHOD_LAUNCH_TEXT_SCREEN:
            rootScreen.present(TextInputViewController.instanceFromStoryboard(),
                               animated: true, completion: nil)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    func sendText(_ text: String) {
        channel.invokeMethod(PlatformText.METHOD_SEND_TEXT, arguments: text)
    }
}
